lib/popcorn/pop_json.nit

```diff
@@ -0,0 +1,172 @@
+#		if form == null then return # Deserialization error: let popcorn return a HTTP 400
+#
+#		# TODO do something with the input
+#		print form.name
+#	end
+# end
+#
+# class MyForm
+#	serialize
+#
+#	var name: String
+# end
+#
+# class MyFormValidator
+#	super ObjectValidator
+#
+#	init do
+#		add new StringField("name", min_size=1, max_size=255)
+#	end
+# end
+# ~~~
+module pop_json
+
+import pop_handlers
+import pop_validation
+
+redef class Handler
+
+	# Validator used to check body input
+	#
+	# Here we use the `pop_validation` module to validate JSON document from the request body.
+	var validator: nullable DocumentValidator = null
+
+	# Validate body input with `validator`
+	#
+	# Try to validate the request body as a json document using `validator`:
+	# * Returns the validated string input if the result of the validation is ok.
+	# * Answers a json error and returns `null` if something went wrong.
+	# * If no `validator` is set, returns the body without validation.
+	#
+	# Example:
+	#
+	# ~~~nit
+	# class ValidatedHandler
+	#	super Handler
+	#
+	#	redef var validator = new MyObjectValidator
+	#
+	#	redef fun post(req, res) do
+	#		var body = validate_body(req, res)
+	#		if body == null then return # Validation error
+	#		# At this point popcorn returned a HTTP 400 code with the validation error
+	#		# if the validation failed.
+	#
+	#		# TODO do something with the input
+	#		print body
+	#	end
+	# end
+	#
+	# class MyObjectValidator
+	#	super ObjectValidator
+	#
+	#	init do
+	#		add new StringField("name", min_size=1, max_size=255)
+	#	end
+	# end
+	# ~~~
+	fun validate_body(req: HttpRequest, res: HttpResponse): nullable String do
+		var body = req.body
+
+		var validator = self.validator
+		if validator == null then return body
+
+		if not validator.validate(body) then
+			# print validator.validation.to_json
+			res.json(validator.validation, 400)
+			return null
+		end
+		return body
+	end
+
+	# Deserialize the request body
+	#
+	# Returns the deserialized request body body or `null` if something went wrong.
+	# If the object cannot be deserialized, answers with a HTTP 400.
+	#
+	# See `BODY` and `new_body_object`.
+	#
+	# Example:
+	# ~~~nit
+	# class MyDeserializedHandler
+	#	super Handler
+	#
+	#	redef type BODY: MyObject
+	#	redef fun new_body_object(d) do return new MyObject.from_deserializer(d)
+	#
+	#	redef fun post(req, res) do
+	#		var form = deserialize_body(req, res)
+	#		if form == null then return # Deserialization error
+	#		# At this point popcorn returned a HTTP 400 code if something was wrong with
+	#		# the deserialization process
+	#
+	#		# TODO do something with the input
+	#		print form.name
+	#	end
+	# end
+	#
+	# class MyObject
+	#	serialize
+	#
+	#	var name: String
+	# end
+	# ~~~
+	fun deserialize_body(req: HttpRequest, res: HttpResponse): nullable BODY do
+		var body = req.body
+		var deserializer = new JsonDeserializer(body)
+		var form = new_body_object(deserializer)
```

Comment:
It seems to me that the deserialization is already working in a loosely typed word since it does not know what it is deserializing.

I think you can drop `new_body_object` completely an rely on the deserializer. You do not need to explicite what class to instantiate. You just need to check that what is returned by the deserializer is acceptable

~~~nit
var body = req.body
var deserializer = new JsonDeserializer(body)
￼var form = deserializer.deserialize(body)
if not form isa BODY or not deserializer.errors.is_empty then
	res.json_error("Bad input", 400)
	return null
end
return form
~~~￼